### ISSUE TITLE
Fix test for test_field_metadata_custom_mapping

### ISSUE THREAD

**Issue by hydrogen602** [RustPython/RustPython#5232]
I'm looking at implementing the changes required for the test `test_field_metadata_custom_mapping` in `Lib/test/test_dataclasses.py` to pass. 

Specifically, accessing a non-existing field in the metadata of a dataclass field returns a `KeyError` instead of an `AttributeError` that CPython 3.12 produces. 

So this fails:
```python
with self.assertRaises(AttributeError):
  fields(C)[0].metadata["b"]
```

I'm currently working on this issue.

**Comment by hydrogen602** [comment]
It looks like the CPython documentation is wrong:

The dataclass docs state that the object given to metadata will be wrapped in a [MappingProxyType()](https://docs.python.org/3/library/types.html#types.MappingProxyType)
https://docs.python.org/3/library/dataclasses.html#dataclasses.field

Following the link, it states that indexing with a key on MappingProxyType throws a `KeyError`!

> proxy[key]
>     Return the item of the underlying mapping with key key. Raises a [KeyError(https://docs.python.org/3/library/exceptions.html#KeyError) if key is not in the underlying mapping.

But CPython produces an AttributeError and CPython's unit tests expect an AttributeError:
https://github.com/python/cpython/blob/5f4c7cf3f49629bcf890c305e0a90ffd708110a2/Lib/test/test_dataclasses/__init__.py#L2065C1-L2066C39

---
In summary, the RustPython implementation follows the documentation but the CPython implementation does not.
